Handle each Eclipse ASTs not more than two times (diet + full)

Rawi01 opened the pull request:
This PR fixes #2489 and may also solve #1717 (at least to some extent, I identified this problem using the instructions mentioned there)

rspilker commented:
Some remarks:
- Typo in `completlyHandled` => `completelyHandled`
- I'm not sure if using a Boolean is the best idea. An enum would possibly be better. But that is because of my experience with auto-unboxing Boolean variables that happened to contain `null`. Then again, this is local code and not API, so I can live with it.
- I do have thread-safety concerns. We know that under some circumstances, several threads in Eclipse concurrently parse the same code. `WeakHashMap` is not inherently thread-safe. So we probably need to either use a synchronized block, or a `ConcurrentHashMap<WeakReference<CompilationUnitDeclaration>, Boolean>`.

Rawi01 commented:
Thanks for the feedback!
I added an enum for the state and added a synchronized map. Using a `ConcurrentHashMap` with `WeakReference` would slowly leak memory because the reference objects will stay in that map forever.
